### PR TITLE
Fix - Add status on project && Delete referenced project

### PR DESCRIPTION
**Pull request by Patricknshimiyimana**
# What does this PR do?
- this PR fixes the issue of deleting referenced projects
- this PR adds a status field on the project
# Description of Task to be completed?
- Before deleting the project delete all userProject records that references it
- Add a status field in the project model and migration

## Requirements

- [ ] Require dependences installation
- [ ] Require .env update
- [ ] Documentation updated

# How should this be manually tested?
> - Clone this repository
> - checkout to this branch `fix-delete-project`
> - install the dependencies `npm install` and 
> - Run migrations `npm run db:migrate` 
> - Create some users or run seeds `npm run db:seed` and then start the server `npm run dev`
> - Go to `http://localhost:$PORT/api/v1/docs` in your browser to access the api documentation
# Any background context you want to provide?

# What are the relevant pivotal tracker stories?

# Screenshots (if appropriate)

# Questions:
- N/A

### CODE REVIEW

File: src/database/relationships/index.js [hunk 1]
```diff
@@ -14,6 +14,19 @@ export const associate = () => {
     otherKey: "userId",
   });
 
+  DB.Project.hasMany(DB.UserProject, {
+    onDelete: "CASCADE",
+    foreignKey: "projectId",
+  });
```
Comment (nicele08) [inline src/database/relationships/index.js]: It is a good practice to add the "as" keyword to provide meaningful aliases for the associations.

Hint: 
```javascript
DB.Project.hasMany(DB.UserProject, {
  onDelete: "CASCADE",
  foreignKey: "projectId",
  as: "userProjects", // Alias for the association
});
```

File: src/database/relationships/index.js [hunk 2]
```diff
@@ -14,6 +14,19 @@ export const associate = () => {
...
+  DB.UserProject.belongsTo(DB.Project, { foreignKey: "projectId" });
+
+  DB.User.hasMany(DB.UserProject, {
+    onDelete: "CASCADE",
+    foreignKey: "userId",
+  });
```
Comment (nicele08) [inline src/database/relationships/index.js]: Same as here; provide meaningful aliases for the associations.

hint:
```javascript
...
as: "project", // Alias for the association
...
```

File: src/database/relationships/index.js [hunk 3]
```diff
@@ -14,6 +14,19 @@ export const associate = () => {
...
+
+  DB.UserProject.belongsTo(DB.User, { foreignKey: "userId" });
```
Comment (nicele08) [inline src/database/relationships/index.js]: Same as here; provide meaningful aliases for the associations.

**Review by nicele08 — APPROVED** [review]
@Patricknshimiyimana Good work. Just check added comments

**Review by Cherubln — DISMISSED** [review]
@Patricknshimiyimana also consider adding the inProgress field on every project model.

**Comment by Patricknshimiyimana** [comment]
> @Patricknshimiyimana also consider adding the inProgress field on every project model.

And that goes with its related actions too!? Like updating it,...

But I thought we had the project `startDate` and `andDate` to track the inProgress projects

**Comment by Cherubln** [comment]
> And that goes with its related actions too!? Like updating it,...
> 
> But I thought we had the project `startDate` and `andDate` to track the inProgress projects

Yeah, but if you think about it, we should let user decide if the project is completed. I think we shouldn't decide that for them. we can only show them that their project is overdue based on end date. and let them decide if it's finished. that's what I suggest.

**Comment by Patricknshimiyimana** [comment]
Added a `status` field which can be `pending/in-progress or completed`
What do you think @Cherubln ???

File: src/database/migrations/20230307201538-create-project.js [hunk 4]
```diff
@@ -31,6 +31,10 @@ module.exports = {
       endDate: {
         type: Sequelize.DATE,
       },
+      status: {
+        type: Sequelize.ENUM("pending", "in-progress", "completed"),
```
Comment (cherubln) [inline src/database/migrations/20230307201538-create-project.js:35]: Thank you @Patricknshimiyimana, but I think only in-progress and completed are enough. unless you think there would be cases where project could be in pending mode.

**Review by Cherubln — CHANGES_REQUESTED** [review]

File: src/database/migrations/20230307201538-create-project.js (hunk 4 above)
Comment (Patricknshimiyimana) [reply src/database/migrations/20230307201538-create-project.js:35]: If you think about it, all projects that have been just added can't be immediately in-progress or completed!
I think there might be a case where a project is added and it is neither in-progress yet nor completed.

what does others think? @denislohan @Nkbtemmy

Comment (cherubln) [reply src/database/migrations/20230307201538-create-project.js:35]: Nice, @Patricknshimiyimana, it is totally understandable I get it, we need all three options. like when the project start date has not been yet reached it should in pending mode. I am approving these changes.

**Review by Cherubln — APPROVED** [review]
LGTM

**Pull request merged**